File: 141_linked_list_cycle.cpp
// 141. Linked List Cycle
// First solved 05/03/26

 struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};
 
class Solution {
    public:
        bool hasCycle(ListNode *head) {

            // fast/slow method: if there's a cycle in the list, eventually
            // fast and slow will be the same node. otherwise if fast reaches
            // null, that means there's no cycle
            ListNode *fast = head;
            ListNode *slow = head;
    
            while(fast && fast->next) {
                fast = fast->next->next;
                slow = slow->next;
                if(fast == slow) {
                    return true;
                }
            }
            return false;
        }
    };

File: linked-list/141_linked_list_cycle.py
# First solved May 3, 2026

from typing import Optional

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # if no cycle, fast will hit NULL, if there is a cycle, eventually slow and fast will be equal.
        fast = head
        slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            if fast == slow:
                return True
                
        return False